### DISCUSSION TITLE
NoClassDefFoundError error on a Netbeans application using OSGi

### DISCUSSION THREAD

**Discussion by YannLeCorse** [apache/netbeans#7601] [General]
Hi,

I'm facing exceptions at the startup of a Netbeans application based on Maven and using OSGi.
This is a sample project to reproduce the issue I'm facing : [my-application.zip](https://github.com/user-attachments/files/16312731/my-application.zip)

- I created a `nbm-application` with OSGi enabled
- I created an OSGi module defining interface `A`
- I created a NBM module containing
  - Interface `B extends A`
  - Singleton class holding an instance of `A`
  - `ModuleInstall` where we inject an instance of `B` to the singleton above
 
The project is failing in the application unit test do to the issue but it's easier to run it in debug to troubleshot. 

At startup, manifests are read and it looks like the OSGi export packages are correctly initialised in the `org.netbeans.ModuleData` as I can see them in `provides`

Then we go through the `org.netbeans.ModuleManager.doEnable()` and it looks like we don't have the same process for OSGi modules and for Netbeans ones.

Netbeans modules are looking for their public packages and register them in the `ProxyClassPackages`
```
"main"
        at org.netbeans.ProxyClassPackages.addCoveredPackages(ProxyClassPackages.java:45)
	at org.netbeans.ProxyClassLoader.addCoveredPackages(ProxyClassLoader.java:94)
	at org.netbeans.JarClassLoader.<init>(JarClassLoader.java:177)
	at org.netbeans.StandardModule$OneModuleClassLoader.<init>(StandardModule.java:609)
	at org.netbeans.StandardModule.createNewClassLoader(StandardModule.java:545)
	at org.netbeans.StandardModule.classLoaderUp(StandardModule.java:533)
	at org.netbeans.ModuleManager.doEnable(ModuleManager.java:1380)
	at org.netbeans.ModuleManager.enable(ModuleManager.java:1350)
	at org.netbeans.ModuleManager.enable(ModuleManager.java:1275)
	at org.netbeans.core.startup.ModuleList.installNew(ModuleList.java:316)
	at org.netbeans.core.startup.ModuleList.trigger(ModuleList.java:252)
	at org.netbeans.core.startup.ModuleSystem.restore(ModuleSystem.java:298)
	at org.netbeans.core.startup.Main.getModuleSystem(Main.java:173)
	at org.netbeans.core.startup.Main.getModuleSystem(Main.java:142)
	at org.netbeans.core.startup.Main.start(Main.java:303)
	at org.netbeans.core.startup.TopThreadGroup.run(TopThreadGroup.java:98)
	at java.lang.Thread.runWith(Thread.java:1596)
	at java.lang.Thread.run(Thread.java:1583)
```

OSGi modules are doing something quite similar but don't register their export packages in the `ProxyClassPackages`
```
"main"
	at org.netbeans.NetigsoHandle.classLoaderUp(NetigsoHandle.java:163)
	at org.netbeans.ModuleManager.netigsoLoaderUp(ModuleManager.java:601)
	at org.netbeans.NetigsoModule.classLoaderUp(NetigsoModule.java:108)
	at org.netbeans.ModuleManager.doEnable(ModuleManager.java:1380)
	at org.netbeans.ModuleManager.enable(ModuleManager.java:1350)
	at org.netbeans.ModuleManager.enable(ModuleManager.java:1275)
	at org.netbeans.core.startup.ModuleList.installNew(ModuleList.java:316)
	at org.netbeans.core.startup.ModuleList.trigger(ModuleList.java:252)
	at org.netbeans.core.startup.ModuleSystem.restore(ModuleSystem.java:298)
	at org.netbeans.core.startup.Main.getModuleSystem(Main.java:173)
	at org.netbeans.core.startup.Main.getModuleSystem(Main.java:142)
	at org.netbeans.core.startup.Main.start(Main.java:303)
	at org.netbeans.core.startup.TopThreadGroup.run(TopThreadGroup.java:98)
	at java.lang.Thread.runWith(Thread.java:1596)
	at java.lang.Thread.run(Thread.java:1583)
```

As a result, when my NBM module is starting, it tries to load the class `A` defined in my OSGi module. 
As export packages were not registered, we don't find them in the `ProxyClassPackages` and the exception is thrown.

From my understanding it looks like there is an issue in the way we are managing OSGi modules but I'm opening this discussion to validate that I got it right.

- Is my understanding correct ? 
- Is there any reason why we don't register the export-packages from OSGi modules when we are processing them ? 

Thanks

**Comment by YannLeCorse** [comment]
@ebarboni, @mbien sorry for tagging you (I don't know who else to annoy :stuck_out_tongue_closed_eyes:) but I'd like to know if it's sound like a bug (and then I'm happy to work on it) or if it's a bad practice on my side. Would you have a quick look and share your thoughts please ?

**Reply by mbien** [reply]
> sorry for tagging you (I don't know who else to annoy 😝)

unfortunately (or luckily?), I never went deep into OSGI. But I do know who else to annoy. cc @jtulach @matthiasblaesing @neilcsmith-net

**Reply by YannLeCorse** [reply]
Hi guys,

From my understandng the issue I'm facing is related to the delayed OSGi modules start. 
Thus when we are trying to find the `validate` method, classpaths haven't been initialised properly and we're facing the `NoClassDefFoundError`.

The scenario seems to be limited to NB modules having a `ModuleInstall` "linked" to classes from an OSGi module.
Is there a way to disable the delayed OSGi module start ?
If not, would it makes sense to ignore the `NoClassDefFoundError` at this stage and let the process continue ? (a bit like what is curently done for modules not implementing the `validate` method)
Going that way, the modules will start as expected but if the validate method was implemented, it'd not be called.

What's your views ? 

Thx
Yann

**Comment by matthiasblaesing** [comment]
Please provide a minimal sample. Guessing does not help.

**Comment by YannLeCorse** [comment]
> Please provide a minimal sample. Guessing does not help.

There is one attached in the initial message

**Reply by matthiasblaesing** [reply]
A sorry - overlooked

**Reply by YannLeCorse** [reply]
Did you managed to have a look ? 
1) I'd like to confirm if there is an issue in the current NB startup process or if it's a bad practice on my side with the ModuleInstall.
2) If it's a NB issue, do you have any guidance on how it should be addressed ? 
3) If it's a NB issue, do you think catching the `NoClassDefFoundError` is acceptable or not at all ?

Thanks ;)

**Reply by neilcsmith-net** [reply]
Have you tried using `@OnStart` instead of the `ModuleInstall`? https://bits.netbeans.org/23/javadoc/org-openide-modules/org/openide/modules/OnStart.html  Does that work around the issue?

**Reply by YannLeCorse** [reply]
It could work on the sample project provided as it's a very simple one to illustrate the class access issues but it doesn't apply in my real scenario where I got hundred of modules because of the parallel execution.

**Reply by neilcsmith-net** [reply]
Does it fail only in cases where you have an inner class that extends from an OSGi provided class?  Can you extract that into another class called from the module install?  Does that fix the issue?

**Reply by YannLeCorse** [reply]
It fails as soon as the `ModuleInstall` class is somehow related to a class provided by the OSGi module (inner class implementation or external one) from my understanding because we'll access all the classes when we do the `getDeclaredMethod("validate")`.
Thus we go through the `ProxyClassLoader.doFindClass` which is not able to locate the ones from OSGi modules

**Reply by neilcsmith-net** [reply]
For example, if in the provided project you moved the body of the `restored` method into another adjacent class, called from the `ModuleInstall::restored`, does it now work OK?

In order to search for (and possibly call) the `validate` method, the JVM is going to need to be able to load the `ModuleInstall` subclass without OSGi classes available.  That appears to be a limitation of the current support that we need to document.

**Reply by YannLeCorse** [reply]
Oh sorry, I got you now... Yes moving the code of the `restored` method to an adjacent class is working because by doing that the `ModuleInstall` is no more related to the OSGi classes and then we can do reflection on it without facing the classloader issue **but** then we have to put in place a delegate pattern for all implemented methods.

It's a bit weird but it workaround the current limitation so thank you for the suggestion 😃 

At the moment such modules are not working, so I guess I'm the very first one trying to do that (or at least reporting it) and I'm happy to change my application if needed but it's pretty easy to pull an OSGi dependency without noticing it if class structure is more complex so I think we should try to enhance the behaviour anyway but please let me know if you prefer to cancel my PR and just document it.

PR is not perfect, I agree but it improves the behaviour as most of the time validate method is not implemented (as it's called very early in the process) and probably not so often in an OSGi dependency context...

Pleas let me know and thanks again for the suggestion (and time). Much appreciated

**Reply by mbien** [reply]
also watch out during `@OnStop` https://github.com/apache/netbeans/pull/7587#issuecomment-2233460269

**Reply by neilcsmith-net** [reply]
From what I can see, changing the `ModuleInstall` in the example from 

```java
public class Installer extends ModuleInstall {

    @Override
    public void restored() {
        MyService.getInstance().setRegistry(new INbRegistry() {
            @Override
            public void doSomethingElse() {
            }

            @Override
            public void doSomething() {
            }
        });
    }

}
```

to 

```java
public class Installer extends ModuleInstall {

    @Override
    public void restored() {
        Handler.restored();
    }

    private static class Handler {

        private static void restored() {
            MyService.getInstance().setRegistry(new INbRegistry() {
                @Override
                public void doSomethingElse() {
                }

                @Override
                public void doSomething() {
                }
            });
        }

    }

}

```

is enough to not fail on load.  Adding a `validate` method also shows that is correctly handled.

**Reply by neilcsmith-net** [reply]
I missed your reply while writing that!  The important problem with the PR is that it will could cause validation to be silently missed?!

**Reply by neilcsmith-net** [reply]
Incidentally, it's hard to tell from your example why you're using `ModuleInstall` at all.  I would note in the docs for ModuleInstall "most modules should not need this".  I assume there is some reason the code in your module install cannot run when loading the services the module provides?

**Reply by YannLeCorse** [reply]
I'm moving an old app with 400+ modules from the old wrapper approach to the OSGi ones. ModuleInstaller were used long time ago (before annotations) as we did checks (license, jdk, etc) in `validate` for few modules + some registration in the `restore` for main of them.

With wrappers, the various ModuleInstall were called in the module dependencies order and classes are registered sequentially so everything was OK. Unfortunately it's not the case with OSGi 😆.
Annotated classes could be used in some places but as they are processed in parallel it doesn't cover all my needs (at least at start) as ordering is important in my scenario.

I'll close the PR and adapt my own code to the current limitation as suggested as I don't have so many situation where the ModuleInstall is having a dependency to OSGi classes but that needs to be documented as it's error prone and not obvious (especially when not implementing the `validate` 😄)

Thanks again for your time looking at this 😉

**Reply by neilcsmith-net** [reply]
It wasn't meant as a comment on `ModuleInstall` vs the similar annotations.   Registering things in a module install seems backwards to the usual way of discovering services through system lookup, layer files, etc.  Still, legacy ... :smile:

**Comment by YannLeCorse** [comment]
It sounds like a NB issue rather than a bad design so I created the pull request https://github.com/apache/netbeans/pull/7624
